### PR TITLE
Add Lambda for SQL User Custom CloudFormation Resource

### PR DESCRIPTION
**Pull request by sureshc**
Add a lambda to our `Lambda` Stack that backs a custom CloudFormation resource to create/update/delete application SQL users. Extracted from #52796 and #53875

## Testing story

Node tests must be executed manually:
``` bash
sql-user % npm test

> sql-user@1.0.0 test
> mocha *.test.js



  SQL User Custom Resource
REQUEST RECEIVED:
 {"RequestType":"Create","ResourceProperties":{"Privileges":["ALL PRIVILEGES"]},"PhysicalResourceId":"someId"}
{}
{}
[ {}, {} ]
    ✔ should handle Create event successfully
REQUEST RECEIVED:
 {"RequestType":"Delete","ResourceProperties":{},"PhysicalResourceId":"someId"}
{}
    ✔ should handle Delete event successfully


  2 passing (22ms)
```

Check Stack Update:
```
export AWS_PROFILE=codeorg-admin
bundle exec rake stack:lambda:validate ADMIN=true

% bundle exec rake stack:lambda:validate ADMIN=true
Finished stack:lambda:environment (less than a minute)
Pending update for stack `lambda`:
Add SQLUserLambda [AWS::Lambda::Function]
Add SQLUserRole [AWS::IAM::Role]
Finished stack:lambda:validate (less than a minute)
```

## Deployment strategy

Deploy manually via an Infrastructure engineer using an admin role with a rake task:

``` bash
export AWS_PROFILE=codeorg-admin
bundle exec rake stack:lambda:start ADMIN=true
```

## Follow-up work

- Change our rake stack `validation` task to execute all node Lambda tests in addition to creating a Stack ChangeSet.
- Add a method to `Cdo::CloudFormation::Lambda` that packages Ruby Lambdas (install gems, zip, upload to S3)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: aws/cloudformation/lambda.yml.erb [hunk 1]
```diff
@@ -273,6 +273,48 @@ Resources:
...
+      Code: <%=package_node_lambda 'sql-user' %>
+      Handler: index.handler
+      Runtime: nodejs18.x
+      MemorySize: 1024
+      Timeout: 30
+      Role: !GetAtt SQLUserRole.Arn
+      VpcConfig:
+        SecurityGroupIds: [!ImportValue VPC-FrontendSecurityGroup]
+        SubnetIds: <%=subnets.to_json%>
+  SQLUserRole:
```
Comment (cat5inthecradle) [inline aws/cloudformation/lambda.yml.erb]: Minor nit on naming, out of context if I was looking at a list of IAM Roles it wouldn't be immediately clear what this is for.

```suggestion
  SQLUserLambdaRole:
```

File: lib/cdo/cloud_formation/lambda.rb [hunk 2]
```diff
@@ -34,6 +34,61 @@ def js_erb(filename, **args)
...
+      Dir.chdir(absolute_directory) do
+        # Use the `ci` parameter to only install the versions identified in the lock file.
+        # Use `--only=prod` to skip dev dependencies.
+        RakeUtils.npm_install 'ci --only=prod'
+      end
+    end
+
+    # Zip a directory containing a Lambda's source code and dependencies, upload to S3, and return the S3 location
+    # to assist in populating the `Code` Property of a CloudFormation template `AWS::Lambda::Function` Resource.
+    # Assumes CloudFormation custom Resource Lambdas are in `/aws/cloudformation/lambdas/`.
```
Comment (cat5inthecradle) [inline lib/cdo/cloud_formation/lambda.rb]: Is this limited to Cloudformation Custom Resources, or would this be fine for any lambda?

**Review by cat5inthecradle — APPROVED** [review]

**Comment by unlox775-code-dot-org** [comment]
Overall, it looks OK to me.

I'm still in the process of reading into the current IAM policies, etc.  There are patterns I don't know like how you guys run these in daily use.  There are some other Lambdas that are allowed to run any other lambda (could be tightened up), which is the only security concern I ran across, as it could theoretically be an escalation path to create a MySQL user.

I'm not sure if there is any way you could use parameter binding instead of string substituting into your SQL queries, to avoid injection.  As it's internal I don't think it's an issue, but more of a standard practice

File: lib/cdo/cloud_formation/lambda.rb (hunk 2 above)
Comment (sureshc) [reply lib/cdo/cloud_formation/lambda.rb]: This is for any Lambda that we need to deploy with our existing rake `stack` tasks (using our `Cdo::CloudFormation::Lambda` module). My hope is that in the long term we would use more modern methods (like we do in javabuilder) to package and deploy lambdas.

Comment (cat5inthecradle) [reply lib/cdo/cloud_formation/lambda.rb]: We could remove the 'custom resource' bit then, as this could be used for any lambda.

Comment (sureshc) [reply lib/cdo/cloud_formation/lambda.rb]: Ah. Good point. I’ll update the comments to note that.

Comment (sureshc) [reply lib/cdo/cloud_formation/lambda.rb]: Good point. Done!

**Comment by sureshc** [comment]
Deployed 2023-10-13 21:00 UTC with:

``` bash
% bundle exec rake stack:lambda:start
GetSecretValue: development/cdo/slack_bot_token
Flushing Cdo::Metrics::Buffer, waiting Infinity seconds
Finished stack:lambda:environment (less than a minute)
Flushing Cdo::Metrics::Buffer, waiting Infinity seconds
Flushing Cdo::Metrics::Buffer, waiting Infinity seconds
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn  --production
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig npm install  ci --only=prod
PKG_CONFIG_PATH=/usr/X11/lib/pkgconfig yarn
Pending update for stack `lambda`:
Add SQLUserLambdaRole [AWS::IAM::Role]
Add SQLUserLambda [AWS::Lambda::Function]
Proceed? [y/n]
y
 Stack update requested, waiting for provisioning to complete...
.2023-10-13 21:13:19 UTC- SQLUserLambdaRole [CREATE_IN_PROGRESS]
2023-10-13 21:13:19 UTC- SQLUserLambdaRole [CREATE_IN_PROGRESS]: Resource creation Initiated
..2023-10-13 21:13:35 UTC- SQLUserLambdaRole [CREATE_COMPLETE]
2023-10-13 21:13:36 UTC- SQLUserLambda [CREATE_IN_PROGRESS]
.2023-10-13 21:13:37 UTC- SQLUserLambda [CREATE_IN_PROGRESS]: Resource creation Initiated
......................................2023-10-13 21:17:22 UTC- SQLUserLambda [CREATE_COMPLETE]
.2023-10-13 21:17:25 UTC- lambda [UPDATE_COMPLETE_CLEANUP_IN_PROGRESS]

Stack update complete.
Finished stack:lambda:start (5 minutes)
```

**Pull request merged**